### PR TITLE
Navigation Component: Merge feature branch

### PR DESCRIPTION
**Pull request by psealock**
## Description

Expose a `<Navigation />` component from `@wordpress/components` to render an array of menu items in a hierarchal fashion. The component is highly composable, accepting a child function that allows full control to the consuming application. It is also free from excessive styling so it can be adapted to many UIs.

This component will be useful in the [Full Site Editing Sidebar project](https://github.com/WordPress/gutenberg/issues/23939) and [WooCommerce's Navigation project](https://cdpn.io/jameskoster/debug/wvKrmwv#).

## How has this been tested?

This has been tested using Storybook and unit tests. In order to test the component, run the following:

```
npm run storybook:dev
```

Next, go to Components > Navigation and see the component in use.

## Screenshots 

Screen Shot 2020-08-31 at 3 33 57 PM
Screen Shot 2020-08-31 at 3 34 05 PM

## Types of changes

New feature: This PR introduces a new component.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Review by Copons — CHANGES_REQUESTED** [review]
I've got two minor complaints and one comment, but otherwise this works as expected!

I think this component can be a solid foundation to be improved upon by integrating style and features suggested, for example, by @ItsJonQ in https://github.com/ItsJonQ/g2/issues/20.

**Comment by ItsJonQ** [comment]
Thanks for your continued efforts on all! I checked out the code and I poked around with the Storybook example.

It looks like the (infinite?) nesting feature is now working :D, which is great! The menu items are tabbable and can be "clicked" with keyboard presses, which is nice.

A couple of observations from my end:


### Needs to render more than just links

One of [FSE's sidebar designs](https://github.com/WordPress/gutenberg/issues/23939#issuecomment-658658405) features a UI where the elements that appear within the menu are **not** just links. Instead it's a searchable list (of pages). This particular view is a nested one (maybe level 3?)

At the moment, it appears that nested level renderings are handled via `id` properties (`id` and `parent`).

Below is a GIF prototyping that interaction:



It comes from a G2 Components storybook prototype:
https://g2-components.xyz/iframe.html?id=components-navigator--sidebar&viewMode=story

FSE appears to be taking advantage of the new Sidebar designs. The `<Navigation />` component needs to be flexible + composable enough to (easily) render various experiences.



https://github.com/WordPress/gutenberg/issues/23939#issuecomment-683049485

#### Sub-Nav Rendering Lifecycles

Continuing with the note above (on making sub-nav content more flexible)... `<Navigation />` should support a way for these sub-navs to have their own lifecycle rendering hooks (e.g. `useEffect`, `useState`, etc...).

For example, the FSE Template mock (above) may require fetching/loading of templates via `useEffect`. Perhaps loading additional templates via a "Load More" interaction, which it needs to keep track of in state (`useState`).

A simpler example would perhaps be async fetching of badge counts.


### (Avoid) Animation on mount

The menu slides on initial mount. Either the `<Animate />` component or the mechanism handling the animations within `<Navigation />` needs to animate when menu item is engaged with (clicked).

---

Hope this feedback helps!

**Comment by Copons** [comment]
> ### Needs to render more than just links

@ItsJonQ It is currently possible to render anything in a list with the `LinkComponent` prop.

Right now you can see it used with a custom Button in the storybook:
```jsx
const CustomRouterLink = ( { children, onClick } ) => {
	return <Button onClick={ onClick }>{ children }</Button>;
};

// ...

const data = {
	title: 'Internal link',
	id: 'item-5',
	LinkComponent: CustomRouterLink,
};
```

Nothing (visual glitches notwhitstanding) stops us from rendering something else in there:

```jsx
const CustomImageLink = ( { onClick } ) => {
	return ( <a onClick={ onClick }>
		<img src="page-template.jpg" />
	</a> );
}
```

---

Regarding the search.
The current implementation renders the menu title as a simple `<h1>`.
We can update it to be its own component that turns into a search field on click, as suggested in https://github.com/WordPress/gutenberg/issues/23939#issuecomment-658658405.

Using that new "SearchTitle" component could be opt-in, so if you don't need it you can keep using `<h1>`.

File: packages/components/src/navigation/stories/index.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	},
+	{
+		title: 'Sub Child 2',
+		id: 'sub-child-2',
+		parent: 'child-3',
+	},
+	{
+		title: 'External link',
+		id: 'item-4',
+		href: 'https://wordpress.com',
```
[the remark this answers is not in the document]
Comment (psealock) [reply packages/components/src/navigation/stories/index.js]: Good idea. Updated in https://github.com/WordPress/gutenberg/pull/24920/commits/5897bfaaee71fbeefc90d5b43f1167b1e63ab00b

File: packages/components/src/navigation/README.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+### `data`
+
+-   Type: `array`
+-   Required: Yes
+
+An array of config objects for each menu item.
+
+Config objects can be represented
+
+#### `config.title`
```
[the remark this answers is not in the document]
Comment (psealock) [reply packages/components/src/navigation/README.md]: Makes sense, thank you. https://github.com/WordPress/gutenberg/pull/24920/commits/5897bfaaee71fbeefc90d5b43f1167b1e63ab00b

File: packages/components/src/navigation/index.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+					</div>
+				) }
+			</Animate>
+		</Root>
+	);
+};
+
+export default Navigation;
+export { default as NavigationMenu } from './menu';
+export { default as NavigationMenuItem } from './menu-item';
```
[the remark this answers is not in the document]
Comment (psealock) [reply packages/components/src/navigation/index.js:116]: Agree. I'm following a pattern already in place though:

https://github.com/WordPress/gutenberg/blob/5897bfaaee71fbeefc90d5b43f1167b1e63ab00b/packages/components/src/tree-grid/index.js#L161-L163

**Comment by psealock** [comment]
Thanks for the review @ItsJonQ !

>Continuing with the note above (on making sub-nav content more flexible)... <Navigation /> should support a way for these sub-navs to have their own lifecycle rendering hooks (e.g. useEffect, useState, etc...).

As noted above, there is the ability to pass in custom components, complete with their own lifecycle hooks.

>For example, the FSE Template mock (above) may require fetching/loading of templates via useEffect. Perhaps loading additional templates via a "Load More" interaction, which it needs to keep track of in state (useState).

Interesting use case, thanks for noting. This would make sense for paginated data. I'm fairly sure the composibility affords this possibility. I'm wondering if a more involved Storybook exploration makes sense here. It may be a good idea for demonstrating a search input as well.

>A simpler example would perhaps be async fetching of badge counts.

Excellent point. Here is an issue to track that task: https://github.com/WordPress/gutenberg/issues/24950

>(Avoid) Animation on mount

Indeed: https://github.com/WordPress/gutenberg/issues/24951

**Comment by Copons** [comment]
Another thing that comes to mind, when thinking about customizability, is that we can extend the `data` items with more properties as needed.

For example, in https://github.com/WordPress/gutenberg/issues/23939#issuecomment-683960667, there are two interesting proposal for the FSE navigation: groups and previews on hover.
Whereas now items have just the essential props, we might want to eventually explore additional properties that would automatically enable more complicated features, without having to "overload" the `LinkComponent` prop.

The simplest example could be having a `type` property that handles what kind of output we want from `NavigationMenuItem`.
Currently, it only either renders a text button, or an entirely custom component.
With a `type` prop, we could provide a set of pre-configured components, so that consumers won't have to figure it out by themselves, leading to inconsistent experiences.

`type` could even be "triggered" by other properties. E.g. having an `image` property with an image URL would force the menu item to be `type: 'image'`, or something like that.

I guess what I have in mind is something similar to the block type definition, with all those `supports`, `example`, etc. properties.

File: packages/components/src/navigation/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Copons) [reply packages/components/src/navigation/index.js:116]: Ha! I've tried opening a bunch of random `index.js` files and never found the same pattern, but I suspected it was there somewhere. 😄

**Comment by psealock** [comment]
Thanks for the thoughtful feedback @Copons !

>Another thing that comes to mind, when thinking about customizability, is that we can extend the data items with more properties as needed.

This is the strategy employed on the WooCommerce side:

https://github.com/woocommerce/navigation/blob/3001848b8798f0f81060d2160adcb9e84692ab4d/client/navigation-container.js#L55-L60

A `menuId` is added to items with `primary|secondary` to give a sectioned experience:

Screen Shot 2020-08-28 at 11 20 59 AM

Because we opted to provide a minimalist solution, it made sense to offload that logic to the consumer. If this is a feature required by FSE as well, it could pay to investigate a solution where sections are dictated in a more formal way, such as via the data array or perhaps something else entirely.

>Whereas now items have just the essential props, we might want to eventually explore additional properties that would automatically enable more complicated features, without having to "overload" the LinkComponent prop.

Interesting idea here. Although the `LinkComponent` prop does afford the ability to pass in something that shows a preview on hover, for example, you are right in that `LinkComponent` could be the over used. In addition to the example you provided, a next step could be to compile use cases to identify redundant ones.

>With a type prop, we could provide a set of pre-configured components, so that consumers won't have to figure it out by themselves, leading to inconsistent experiences.

Nice! This exploration would lend itself well to a follow up PR

**Comment by Copons** [comment]
I've noticed a bit of a odd behaviour with e652c32. It might not be a big deal, but let me point it out anyway.

The Back button doesn't update the active item.

It's noticeable with the new "non-navigation link" which only shows up if the active item is not `child-2`.
Clicking it will navigate inside the sub-menu, and activate the sub-menu item `child-2`; the link correctly disappears.
Then, going back to the root menu, the active item remains `child-2` (which is not part of the current menu), and the non-navigation link doesn't show up until we manually click/activate a menu item.

I wonder if it would make sense for the Back button to also clear the active item.

E.g.
```jsx
// stories/index.js
const [ active, setActive ] = useState( 'item-1' );

return (
	<NavigationBackButton setActive={ setActive }>
		<Icon icon={ arrowLeft } />
		{ parentLevel.title }
	</NavigationBackButton>
);

// navigation/index.js
const NavigationBackButton = ( { children: backButtonChildren, setActive } ) => {
	if ( ! parentLevel ) {
		return null;
	}

	const onClick = () => {
		setActiveLevelId( parentLevel.id );
		setActive( undefined );
	};

	return (
		<Button
			className="components-navigation__back-button"
			isPrimary
			onClick={ onClick }
		>
			{ backButtonChildren }
		</Button>
	);
};
```

**Comment by ItsJonQ** [comment]
Haii. Wanted to give a heads up. It looks like Global Styles designs is exploring the potential of having nested navigation as well, but for the right sidebar:

Screen Shot 2020-09-01 at 5 10 21 PM

Figma: https://www.figma.com/file/oEkcAyhIvPFMVEAO8EImvA/Global-Styles?node-id=421%3A4508

I did an early prototype here:
https://g2-components.xyz/iframe.html?id=examples-wip-globalstylessidebar--default&viewMode=story

Ideally, this Navigation component/system should be flexible enough to create this right sidebar/nav experience as well, since the mechanics are very similar.

**Comment by ItsJonQ** [comment]
> I'm wondering if a more involved Storybook exploration makes sense here. It may be a good idea for demonstrating a search input as well.

@psealock It would be awesome to see this solution tackling more complex cases 💪 .
Storybook prototypes would be a great place to test and see.

Inspiration/use-case examples may be tricky to draw from as FSE + Global Styles designs are still in flux.

**Comment by psealock** [comment]
@Copons 

>Then, going back to the root menu, the active item remains child-2 (which is not part of the current menu), and the non-navigation link doesn't show up until we manually click/activate a menu item.

This is by design. Previous designs called for navigation to a set of child items (forward or backward) to set the active item as the first item, but this has been abandoned so that users can navigate the navigation without actually changing the page. Only when an item is clicked does the active item change and page navigation occurs.

>Then, going back to the root menu, the active item remains child-2 (which is not part of the current menu), and the non-navigation link doesn't show up until we manually click/activate a menu item.

Right, another way to look at it is you'd be on the "Child 2" page already, so the link wouldn't be there until you navigated away.

File: packages/components/src/navigation/index.js [hunk 4]
```diff
@@ -67,6 +67,14 @@ const Navigation = ( { activeItemId, children, data, rootTitle } ) => {
 		}
 	}, [ activeItem ] );
 
+	const isMounted = useRef( false );
```
Comment (Copons) [inline packages/components/src/navigation/index.js:70]: I think this doesn't need an initial value, e.g. `const isMounted = useRef();` should work just as well.

**Review by Copons — APPROVED** [review]
Nitpick aside, this looks and works fine to me!

I'm proposing to merge it, and proceed by iterating on it in smaller standalone PRs.
Working in feature branches will only make this much more complicated to review eventually.

We might also be more clear that this is a _really_ experimental component.
I'm not sure if there's an established approach for this, but maybe we could add a note in the readme, or something like that. 🤔

**Comment by afercia** [comment]
A quick note on the "sliding sidebars" pattern. This pattern is not new in WordPress and it's, more or less, like the Customizer sidebars. Those are known to be an accessibility anti-pattern and are a terrible experience for keyboard users and assistive technologies users. When a new sidebar "slides in", the previous content disappears producing a loss of context that's hard to understand for users, especially screen readers users.

This was also quickly discussed during the latest weekly accessibility meeting and probably the team will need to discuss it again bt I'm pretty sure the team has big concerns on the usage of this pattern, being also, as said, a known a11y issue.

**Comment by Copons** [comment]
> A quick note on the "sliding sidebars" pattern. This pattern is not new in WordPress and it's, more or less, like the Customizer sidebars. Those are known to be an accessibility anti-pattern and are a terrible experience for keyboard users and assistive technologies users. When a new sidebar "slides in", the previous content disappears producing a loss of context that's hard to understand for users, especially screen readers users.
> 
> This was also quickly discussed during the latest weekly accessibility meeting and probably the team will need to discuss it again bt I'm pretty sure the team has big concerns on the usage of this pattern, being also, as said, a known a11y issue.

@afercia Thanks for chiming in!
I was thinking about a11y in my experimental proposal (#25057), which has a very different data handling than this PR, but its user-facing behaviour is roughly the same.

When the user navigates from a level to another, the former level un-renders and the latter renders.
In other words, the previous level is completely replaced by the next.

Would it make sense if on level change we automatically focused the first item of the new level?
Or, if it's a nested level, the back button (which is positioned above the list, so it would have the first tabindex by default)?

Should we instead try to work with always-rendered nested lists, and only toggle their visibility (which is kinda complicated, but that's not the point here 🙂)?

**Comment by afercia** [comment]
@Copons thanks for the ping.

> When the user navigates from a level to another, the former level un-renders and the latter renders.
> In other words, the previous level is completely replaced by the next.

I'm not sure that makes a difference. Whether a part of a UI that is currently used by users gets hidden or removed from the DOM, that's a big accessibility problem. There's a complete loss of context. Information on the number of items in the list and their nesting level is broken. Screen reader users wouldn't have a clue that a part of the UI just doesn't exist any longer (why should they?). Moving focus programmatically should be rare and only implemented when it's the expected behavior for [some ARIA pattern](https://www.w3.org/TR/wai-aria-practices-1.1/). In my years of experience in contributing to the accessibility team I can say with no doubt that the Customizer sidebar is one of the patterns we received a lot of complaints about. Also, I would like to note that in other parallel issues about sidebars, there are ongoing explorations and discussions on how to move away from the concept of sidebars.

**Comment by Copons** [comment]
Closing as replaced by #25057.

Thank you so much @psealock and @joshuatf for working on this!
The other approach couldn't have been conceived without the excellent concept and groundwork laid out here — not to mention unit tests and other bits and pieces that I always forget to add!

I've also especially appreciated how gracefully y'all reacted to me casually showing up out of the blue and trampling all over the place. 🙇‍♂️

**Pull request closed without merging**